Add missing test for CallerArgumentExpression

333fred opened the pull request:
Closes https://github.com/dotnet/roslyn/issues/52745.

333fred commented:
This test, combined with https://github.com/dotnet/csharplang/pull/5438 to update the spec with the conversion rules, closes out the test plan for CallerArgumentExpression. @dotnet/roslyn-compiler for a very small test-only change.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:
```diff
@@ -5752,5 +5752,28 @@ static void Main()
+", targetFramework: TargetFramework.NetCoreApp);
+
+            comp.VerifyDiagnostics(
+                // (5,16): error CS8964: The CallerArgumentExpressionAttribute may only be applied to parameters with default values
```
>may only be applied to parameters with default values [](http://example.com/codeflow?start=79&length=53)

It feels like we also need to test scenario that satisfies this requirement, even if that would be an error on its own.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:
```diff
@@ -5752,5 +5752,28 @@ static void Main()
+        [InlineData("out")]
+        [InlineData("ref")]
+        [InlineData("in")]
+        public void CallerArgumentExpression_OnRefParameter(string refType)
```
>CallerArgumentExpression_OnRefParameter [](http://example.com/codeflow?start=20&length=39)

I think we need to make sure we are testing the same scenario for VB.

AlekseyTs commented:
Done with review pass (commit 1)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [971899219](https://github.com/dotnet/roslyn/pull/57805#issuecomment-971899219) [](http://example.com/codeflow?ancestors=971899219,971899219)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:
```diff
@@ -5752,5 +5752,71 @@ static void Main()
+                Diagnostic(ErrorCode.ERR_BadCallerArgumentExpressionParamWithoutDefaultValue, "CallerArgumentExpression").WithLocation(5, 16),
+                // (5,47): error CS1741: A ref or out parameter cannot have a default value
+                // void M(int i, [CallerArgumentExpression("i")] out string s = null)
+                Diagnostic(ErrorCode.ERR_RefOutDefaultValue, "out").WithLocation(5, 47)
```
>"out" [](http://example.com/codeflow?start=61&length=5)

Should this be "ref" for ref?

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs:
```diff
@@ -5752,5 +5752,71 @@ static void Main()
+{
+    Console.WriteLine(s);
+}
+", targetFramework: TargetFramework.NetCoreApp);
```
>targetFramework: TargetFramework.NetCoreApp [](http://example.com/codeflow?start=3&length=43)

I guess default target framework would work just fine since the test is for CoreClrOnly.

333fred replied:
No, it will still compile against Netstandard2.0 APIs, which doesn't have `CallerArgumentExpression`.

AlekseyTs reviewed (approved):
LGTM (commit 3), assuming CI is passing
